synapse/handlers/stats.py

```diff
@@ -111,6 +111,12 @@ def _unsafe_process(self):
                 if self.pos is not None:
                     yield self.store.update_stats_positions(self.pos)
 
+        # Then count deltas for total_events and total_event_bytes.
+        with Measure(self.clock, "stats_total_events_and_bytes"):
+            self.pos = yield self.store.incremental_update_total_events_and_bytes(
+                self.pos
+            )
```

Comment:
This needs to be in the loop and happen before we do the `update_stats_positions`, no? And then it should be given the range it should be working on?

Reply:
`incremental_update_total_events_and_bytes` will update the relevant positions itself, transactionally. Note that it is a separate process from the state deltas (not all events will cause a state delta; only state events).

I don't see why it needs to be in the loop.

Reply:
1. Note that on busy servers we may never exit the loop, as more stuff comes in during processing.
2. Does this mean the values in the stats table are not consistent with each other?

Reply:
1. In that case, it may be desirable to have it both in and out of the loop…? The dual issue is that quiet servers may not see state events for days in normal operation, even though normal events are flying through frequently.
2. Hmm, how do you mean?

Reply:
So right now (mostly) when you look at the stats tables all the columns will be consistent, whereas if you do the total event calculation separately that is no longer true. 

Its worth noting that `notify_new_event` gets called everytime there is a new event, not just new state, so your loop can be something like:

```python
while True:
    deltas = yield get_deltas(pos)
    if deltas:
        handle_deltas(deltas)

    counts = get_events(pos)
    if counts:
        handle_counts(counts)

    if not deltas and not counts:
        break
```

Reply:
> So right now (mostly) when you look at the stats tables all the columns will be consistent, whereas if you do the total event calculation separately that is no longer true.

That's true, but if we keep the updates close together in time, I hope that shouldn't be an issue.

Good point about the pseudocode. How to handle the `with Measure`? Wrap each section in its own (does it matter that it is called more often?)?

e.g.
```python
while True:
    with Measure(...):
        deltas = yield get_deltas(pos)
        if deltas:
            handle_deltas(deltas)

    with Measure(...):
        counts = get_events(pos)
        if counts:
            handle_counts(counts)

    if not deltas and not counts:
        break
```

Reply:
It's fine for measure blocks to be called frequently :+1: